### PR TITLE
feat(TDI-43823) : Uses of common-stream-io

### PR DESCRIPTION
**Pull request by ypiel-talend**
https://jira.talendforge.org/browse/TDI-43823

The connector used to return some Pojo or jsonObject. Now it returns Records since it uses common-stream-io.
The user can now choose if he wants the answer body as json (parsed) or raw text.

### CODE REVIEW

File: rest/src/main/java/org/talend/components/rest/service/RestService.java [hunk 1]
```diff
@@ -136,11 +127,11 @@
...
-    private Response<byte[]> call(final RequestConfig config, final Map<String, String> headers,
+    private Response<InputStream> call(final RequestConfig config, final Map<String, String> headers,
             final Map<String, String> queryParams, final Body body, final String surl,
             final RedirectContext previousRedirectContext) {
 
-        Response<byte[]> resp = null;
+        Response<InputStream> resp = null;
 
         log.info(i18n.request(config.getDataset().getMethodType().name(), surl,
                 config.getDataset().getDatastore().getAuthentication().getType().toString()));
```
Comment (clesaec) [inline rest/src/main/java/org/talend/components/rest/service/RestService.java]: Add a final variable for "config.getDataset().getDatastore().getAuthentication()" and replace in code below.

File: rest/src/test/java/org/talend/components/rest/source/PollingTest.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+// @Environment
+
+/*
+ * @Environment(DirectRunnerEnvironment.class) // Direct runner not necessary since already SparkRunner
+ *
+ * @EnvironmentConfiguration(environment = "Direct", systemProperties = {
+ *
+ * @EnvironmentConfiguration.Property(key = "talend.beam.job.runner", value = "org.apache.beam.runners.direct.DirectRunner")
+ * })
+ */
```
Comment (Voievidko) [inline rest/src/test/java/org/talend/components/rest/source/PollingTest.java]: Remove commented lines or format it somehow.

File: rest/src/main/java/org/talend/components/rest/service/RecordBuilderService.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+    @Service
+    private RecordIORepository ioRepository;
+
+    @Service
+    private RecordBuilderFactory recordBuilderFactory;
+
+    public Iterator<Record> buildFixedRecord(final Response<InputStream> resp, final boolean isCompletePayload,
+            final Format format) {
+        Map<String, String> headers = Optional.ofNullable(resp.headers()).orElseGet(Collections::emptyMap).entrySet().stream()
```
Comment (mbasiuk-talend) [inline rest/src/main/java/org/talend/components/rest/service/RecordBuilderService.java]: May be moved to separate method.

File: rest/src/main/java/org/talend/components/rest/service/RecordBuilderService.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+                new ExtendedRawTextConfiguration(encoding, isCompletePayload));
+
+        final List<Record> headerRecords = headers.entrySet().stream().map(this::convertHeadersToRecords)
+                .collect(Collectors.toList());
+
+        return new IteratorMap<Record, Record>(reader.read(resp.body()),
+                r -> this.buildRecord(r, resp.status(), headerRecords, isCompletePayload, format), true);
+    }
+
+    private ContentFormat findFormat(final Map<String, String> headers, final Format format) {
```
[the remark this answers is not in the document]
Comment (ypiel-talend) [reply rest/src/main/java/org/talend/components/rest/service/RecordBuilderService.java]: good catch

File: rest/src/main/java/org/talend/components/rest/service/client/ContentType.java [hunk 5]
```diff
@@ -14,22 +14,22 @@
...
 public class ContentType {
 
-    final static String DEFAULT_ENCODING = System.getProperty("org.talend.components.rest.default_encoding");
+    final static String DEFAULT_ENCODING = System.getProperty("org.talend.components.rest.default_encoding",
+            StandardCharsets.UTF_8.name());
 
     public final static String HEADER_KEY = "Content-Type";
 
-    public final static String CHARSET_KEY = "charset=";
+    public final static String DEFAULT_CONTENT_TYPE = "text/plain";
```
[the remark this answers is not in the document]
Comment (ypiel-talend) [reply rest/src/main/java/org/talend/components/rest/service/client/ContentType.java]: was linked with previous message, thanks

File: rest/src/main/resources/org/talend/components/rest/configuration/Messages.properties [hunk 6]
```diff
@@ -61,3 +61,8 @@ RequestBody.jsonValue._placeholder = {}
 RequestBody.xmlValue._displayName = Xml
 RequestBody.xmlValue._placeholder = <xml></xml>
 RequestBody.params._displayName = Body parameters
+#
```
[the remark this answers is not in the document]
Comment (ypiel-talend) [reply rest/src/main/resources/org/talend/components/rest/configuration/Messages.properties]: just separators

File: rest/src/test/java/org/talend/components/rest/source/PollingTest.java (hunk 2 above)
Comment (ypiel-talend) [reply rest/src/test/java/org/talend/components/rest/source/PollingTest.java]: removed

File: rest/src/main/resources/org/talend/components/rest/service/Messages.properties [hunk 7]
```diff
@@ -1,22 +1,20 @@
-org.talend.components.rest.service.I18n.healthCheckStatus = Health check status to "{0}" : "{1}"
-org.talend.components.rest.service.I18n.healthChecException = Health check has generated an exception: "{0}"
+org.talend.components.rest.service.I18n.healthCheckStatus = Health check status to {0} : {1}
+org.talend.components.rest.service.I18n.healthCheckException = Health check has generated an exception: {0}
 org.talend.components.rest.service.I18n.healthCheckOk = Connection Successful
-org.talend.components.rest.service.I18n.healthCheckFailed = Can''t access given URL "{0}"
-org.talend.components.rest.service.I18n.malformedURL = Given URL "{0}" is malformed : "{1}"
-org.talend.components.rest.service.I18n.badRequestBody = Given body is not supported: "{0}"
+org.talend.components.rest.service.I18n.healthCheckFailed = Can''t access given URL {0}"
```
[the remark this answers is not in the document]
Comment (ypiel-talend) [reply rest/src/main/resources/org/talend/components/rest/service/Messages.properties]: For escaping, it was ok before. I'll check.

[the remark this answers is not in the document]
Comment (ypiel-talend) [reply rest/src/main/resources/org/talend/components/rest/service/Messages.properties]: Labels used with @Internationalized interface must be escaped
The render is ok
image

File: common-stream-io/stream-rawtext/src/main/java/org/talend/components/common/stream/input/rawtext/RawTextRecordReader.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        this.extendedConfiguration = extendedConfiguration;
+    }
+
+    @Override
+    public Iterator<Record> read(InputStream reader) {
+        if (reader == null) {
+            if (this.extendedConfiguration.isForceOneRow()) {
+                return Collections.singletonList((Record) null).iterator();
+            } else {
+                List<Record> l = Collections.emptyList();
```
[the remark this answers is not in the document]
Comment (ypiel-talend) [reply common-stream-io/stream-rawtext/src/main/java/org/talend/components/common/stream/input/rawtext/RawTextRecordReader.java]: No I can't du to generic and empyList() method. If I try I have got this error in Intellij
image

File: common-stream-io/stream-rawtext/src/main/java/org/talend/components/common/stream/input/rawtext/RawTextRecordReader.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                return l.iterator();
+            }
+        }
+
+        final ByteArrayOutputStream out = new ByteArrayOutputStream();
+        try {
+            final byte[] tt = new byte[512];
```
[the remark this answers is not in the document]
Comment (ypiel-talend) [reply common-stream-io/stream-rawtext/src/main/java/org/talend/components/common/stream/input/rawtext/RawTextRecordReader.java]: yes, buffer is more understable

File: common-stream-io/stream-rawtext/src/main/java/org/talend/components/common/stream/input/rawtext/RawTextRecordReader.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            int nbe = reader.read(tt);
+            while (nbe > 0) {
+                out.write(tt, 0, nbe);
+                nbe = reader.read(tt);
+            }
+        } catch (IOException e) {
```
[the remark this answers is not in the document]
Comment (ypiel-talend) [reply common-stream-io/stream-rawtext/src/main/java/org/talend/components/common/stream/input/rawtext/RawTextRecordReader.java]: e is ok

File: common-stream-io/stream-rawtext/src/main/java/org/talend/components/common/stream/input/rawtext/RawTextRecordReader.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+            } else {
+                List<Record> l = Collections.emptyList();
+                return l.iterator();
+            }
+        }
+
+        final ByteArrayOutputStream out = new ByteArrayOutputStream();
+        try {
+            final byte[] tt = new byte[512];
+            int nbe = reader.read(tt);
```
[the remark this answers is not in the document]
Comment (ypiel-talend) [reply common-stream-io/stream-rawtext/src/main/java/org/talend/components/common/stream/input/rawtext/RawTextRecordReader.java]: done

File: common/src/test/java/org/talend/components/common/collections/IteratorMapTest.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.List;
+
+class IteratorMapTest {
+
+    @Test
+    void noNull() {
+        final List<String> strings = Arrays.asList("Hello", "all", "of", "you");
+        final IteratorMap<String, Integer> it = new IteratorMap<>(strings.iterator(), String::length);
+
+        Assertions.assertEquals(5, it.next());
```
[the remark this answers is not in the document]
Comment (ypiel-talend) [reply common/src/test/java/org/talend/components/common/collections/IteratorMapTest.java]: Same code style that other test in this package.

File: rest/pom.xml [hunk 13]
```diff
@@ -18,15 +18,33 @@
         <!-- Currently I can't execute httpbin tests on CI -->
         <!-- To run ITs test give the following option to mvn : -Dtest.excluded="" -->
         <test.excluded>ITs</test.excluded>
+        <talend.scan.scannedDependencies>org.talend.components:stream-api</talend.scan.scannedDependencies>
+        <!-- I have removed some i18n keys and I have validation errors since it uses _en for validation -->
```
[the remark this answers is not in the document]
Comment (ypiel-talend) [reply rest/pom.xml]: changed

File: rest/src/main/java/org/talend/components/rest/service/RecordBuilderService.java (hunk 3 above)
Comment (ypiel-talend) [reply rest/src/main/java/org/talend/components/rest/service/RecordBuilderService.java]: not really needed

File: rest/src/main/java/org/talend/components/rest/service/RestService.java (hunk 1 above)
Comment (ypiel-talend) [reply rest/src/main/java/org/talend/components/rest/service/RestService.java]: Done

File: common-stream-io/stream-rawtext/src/main/java/org/talend/components/common/stream/input/rawtext/RawTextRecordReader.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+            while (nbe > 0) {
+                out.write(tt, 0, nbe);
+                nbe = reader.read(tt);
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException("Can't build raw text record.", e);
```
[the remark this answers is not in the document]
Comment (ypiel-talend) [reply common-stream-io/stream-rawtext/src/main/java/org/talend/components/common/stream/input/rawtext/RawTextRecordReader.java]: common-stream-io is not yet i18n

File: rest/src/main/resources/org/talend/components/rest/service/Messages.properties (hunk 7 above)
[the remark this answers is not in the document]
Comment (Voievidko) [reply rest/src/main/resources/org/talend/components/rest/service/Messages.properties]: Didn't know it.

File: common-stream-io/stream-rawtext/src/main/java/org/talend/components/common/stream/input/rawtext/RawTextRecordReader.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mbasiuk-talend) [reply common-stream-io/stream-rawtext/src/main/java/org/talend/components/common/stream/input/rawtext/RawTextRecordReader.java]: What about **Collections.emptyIterator()**?

File: rest/src/main/java/org/talend/components/rest/service/RestService.java [hunk 15]
```diff
@@ -136,39 +128,40 @@
...
                 try {
                     URL url = new URL(surl);
                     DigestAuthService das = new DigestAuthService();
                     DigestAuthContext context = new DigestAuthContext(url.getPath(), config.getDataset().getMethodType().name(),
                             url.getHost(), url.getPort(), body == null ? null : body.getContent(),
-                            new UserNamePassword(config.getDataset().getDatastore().getAuthentication().getBasic().getUsername(),
-                                    config.getDataset().getDatastore().getAuthentication().getBasic().getPassword()));
+                            new UserNamePassword(authentication.getBasic().getUsername(),
+                                    authentication.getBasic().getPassword()));
                     resp = das.call(context, () -> client.executeWithDigestAuth(i18n, context, config, client,
```
Comment (clesaec) [inline rest/src/main/java/org/talend/components/rest/service/RestService.java]: Does [Client interface](https://github.com/Talend/connectors-se/blob/ypiel/TDI-43823_REST_use_common-stream-io/rest/src/main/java/org/talend/components/rest/service/client/Client.java) really need several execute methods (executeWithBasicAuth ...) as [RestConfigurer class](https://github.com/Talend/connectors-se/blob/ypiel/TDI-43823_REST_use_common-stream-io/rest/src/main/java/org/talend/components/rest/service/client/RestConfigurer.java) has also access to authentication values (via datastore) and can put all needed values to headers ? just asking ?

Comment (ypiel-talend) [reply rest/src/main/java/org/talend/components/rest/service/RestService.java]: View together, it is ok. This redundancy is due to digest authent which need several calls.

**Comment by Voievidko** [comment]
LGTM

**Pull request merged**